File: apps/reading/src/api/twinSubstrateCrossAssetMergeCompose.ts
/**
 * Twin substrate cross-asset merge compose (pure).
 *
 * Operator vision: every information asset has a twin of insights/questions;
 * that substrate can be merged, referenced, and leveraged across combining
 * contexts on the infinite information platform.
 *
 * merge_executed always false.
 * twin_written always false.
 * store_mutated always false.
 */

export interface TwinSubstrateSlice {
  parent_asset_id: string;
  twin_asset_id?: string;
  /** Caller-supplied insights only. */
  insights: string[];
  /** Caller-supplied questions only. */
  questions: string[];
}

export interface TwinSubstrateCrossAssetMergeInput {
  /** Operator-facing pack id for this merge proposal. */
  pack_id: string;
  slices: TwinSubstrateSlice[];
  operator_ack: boolean;
}

export interface TwinSubstrateCrossAssetMergeCompose {
  pack_id: string;
  parent_asset_ids: string[];
  parent_count: number;
  insight_count: number;
  question_count: number;
  /** Flattened caller-supplied insights (no invent). */
  insights: string[];
  /** Flattened caller-supplied questions (no invent). */
  questions: string[];
  /**
   * True when ≥2 parents, ≥1 insight or question total, operator_ack.
   * Still does not write twins or merge assets.
   */
  merge_ready: boolean;
  /** Always false — pure layer never executes twin merge. */
  merge_executed: false;
  /** Always false — pure layer never writes twin documents. */
  twin_written: false;
  /** Always false — no store mutation. */
  store_mutated: false;
  notes: string[];
  authority: "twin_substrate_cross_asset_merge_compose_advisory";
}

function requireNonEmpty(value: unknown, name: string): string {
  if (typeof value !== "string" || !value.trim()) {
    throw new Error(`${name} must be a non-empty string`);
  }
  return value.trim();
}

/**
 * Propose merging twin substrate across ≥2 parent assets.
 * Never invents insights/questions; never writes or merges.
 */
export function composeTwinSubstrateCrossAssetMerge(
  input: TwinSubstrateCrossAssetMergeInput,
): TwinSubstrateCrossAssetMergeCompose {
  if (!input || typeof input !== "object") {
    throw new Error("input must be an object");
  }
  if (typeof input.operator_ack !== "boolean") {
    throw new Error("operator_ack must be an explicit boolean");
  }
  const pack_id = requireNonEmpty(input.pack_id, "pack_id");
  if (!Array.isArray(input.slices) || input.slices.length < 2) {
    throw new Error("slices must be an array of at least 2 parent twins");
  }

  const notes: string[] = [
    "merge_executed=false — cross-asset twin merge is intent only",
    "twin_written=false — twin documents not created/updated here",
    "store_mutated=false",
    "insights/questions are caller-supplied only (no invent)",
  ];

  const parent_asset_ids: string[] = [];
  const seenParents = new Set<string>();
  const insights: string[] = [];
  const questions: string[] = [];

  for (let i = 0; i < input.slices.length; i++) {
    const sl = input.slices[i];
    if (!sl || typeof sl !== "object") {
      throw new Error(`slices[${i}] must be an object`);
    }
    const parent = requireNonEmpty(
      sl.parent_asset_id,
      `slices[${i}].parent_asset_id`,
    );
    if (seenParents.has(parent)) {
      throw new Error(`duplicate parent_asset_id: ${parent}`);
    }
    seenParents.add(parent);
    parent_asset_ids.push(parent);

    if (sl.twin_asset_id != null) {
      requireNonEmpty(sl.twin_asset_id, `slices[${i}].twin_asset_id`);
    }
    if (!Array.isArray(sl.insights)) {
      throw new Error(`slices[${i}].insights must be an array`);
    }
    if (!Array.isArray(sl.questions)) {
      throw new Error(`slices[${i}].questions must be an array`);
    }
    for (let j = 0; j < sl.insights.length; j++) {
      const ins = requireNonEmpty(
        sl.insights[j],
        `slices[${i}].insights[${j}]`,
      );
      insights.push(ins);
    }
    for (let j = 0; j < sl.questions.length; j++) {
      const q = requireNonEmpty(
        sl.questions[j],
        `slices[${i}].questions[${j}]`,
      );
      questions.push(q);
    }
  }

  const parent_count = parent_asset_ids.length;
  const insight_count = insights.length;
  const question_count = questions.length;
  notes.push(
    `parents=${parent_count} · insights=${insight_count} · questions=${question_count}`,
  );

  const hasSubstrate = insight_count + question_count >= 1;
  const merge_ready =
    input.operator_ack && parent_count >= 2 && hasSubstrate;

  if (!input.operator_ack) {
    notes.push("merge_ready=false — operator_ack required");
  } else if (!hasSubstrate) {
    notes.push(
      "merge_ready=false — no insights/questions (no invent substrate)",
    );
  } else {
    notes.push("merge_ready=true — provisional cross-asset twin pack");
  }

  notes.push("merge_executed=false");
  notes.push("twin_written=false");
  notes.push("store_mutated=false");

  return {
    pack_id,
    parent_asset_ids,
    parent_count,
    insight_count,
    question_count,
    insights,
    questions,
    merge_ready,
    merge_executed: false,
    twin_written: false,
    store_mutated: false,
    notes,
    authority: "twin_substrate_cross_asset_merge_compose_advisory",
  };
}

export function formatTwinSubstrateCrossAssetMergeSummary(
  c: TwinSubstrateCrossAssetMergeCompose,
): string {
  return (
    `merge_ready=${c.merge_ready} · parents=${c.parent_count} · ` +
    `insights=${c.insight_count} · questions=${c.question_count} · ` +
    `merge_executed=false · twin_written=false`
  );
}

File: substrate/twin_substrate_cross_asset_merge_compose.py
"""Twin substrate cross-asset merge compose (pure).

merge_executed, twin_written, store_mutated always False.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any


class TwinSubstrateCrossAssetMergeComposeError(ValueError):
    """Fail-closed validation for twin cross-asset merge."""


@dataclass(frozen=True)
class TwinSubstrateCrossAssetMergeCompose:
    pack_id: str
    parent_asset_ids: tuple[str, ...]
    parent_count: int
    insight_count: int
    question_count: int
    insights: tuple[str, ...]
    questions: tuple[str, ...]
    merge_ready: bool
    merge_executed: bool
    twin_written: bool
    store_mutated: bool
    notes: tuple[str, ...]
    authority: str

    def to_dict(self) -> dict[str, Any]:
        return {
            "pack_id": self.pack_id,
            "parent_asset_ids": list(self.parent_asset_ids),
            "parent_count": self.parent_count,
            "insight_count": self.insight_count,
            "question_count": self.question_count,
            "insights": list(self.insights),
            "questions": list(self.questions),
            "merge_ready": self.merge_ready,
            "merge_executed": False,
            "twin_written": False,
            "store_mutated": False,
            "notes": list(self.notes),
            "authority": "twin_substrate_cross_asset_merge_compose_advisory",
        }


def _require_nonempty(value: object, *, field: str) -> str:
    if not isinstance(value, str) or not value.strip():
        raise TwinSubstrateCrossAssetMergeComposeError(
            f"{field} must be a non-empty string"
        )
    return value.strip()


def compose_twin_substrate_cross_asset_merge(
    *,
    pack_id: object,
    slices: object,
    operator_ack: object,
) -> TwinSubstrateCrossAssetMergeCompose:
    """Propose merging twin substrate across ≥2 parents. Never writes/merges."""
    if not isinstance(operator_ack, bool):
        raise TwinSubstrateCrossAssetMergeComposeError(
            "operator_ack must be an explicit boolean"
        )
    pid = _require_nonempty(pack_id, field="pack_id")
    if not isinstance(slices, list) or len(slices) < 2:
        raise TwinSubstrateCrossAssetMergeComposeError(
            "slices must be an array of at least 2 parent twins"
        )

    notes: list[str] = [
        "merge_executed=false — cross-asset twin merge is intent only",
        "twin_written=false — twin documents not created/updated here",
        "store_mutated=false",
        "insights/questions are caller-supplied only (no invent)",
    ]

    parent_asset_ids: list[str] = []
    seen: set[str] = set()
    insights: list[str] = []
    questions: list[str] = []

    for i, sl in enumerate(slices):
        if not isinstance(sl, dict):
            raise TwinSubstrateCrossAssetMergeComposeError(
                f"slices[{i}] must be an object"
            )
        parent = _require_nonempty(
            sl.get("parent_asset_id"), field=f"slices[{i}].parent_asset_id"
        )
        if parent in seen:
            raise TwinSubstrateCrossAssetMergeComposeError(
                f"duplicate parent_asset_id: {parent}"
            )
        seen.add(parent)
        parent_asset_ids.append(parent)
        if sl.get("twin_asset_id") is not None:
            _require_nonempty(
                sl.get("twin_asset_id"), field=f"slices[{i}].twin_asset_id"
            )
        ins = sl.get("insights")
        qs = sl.get("questions")
        if not isinstance(ins, list):
            raise TwinSubstrateCrossAssetMergeComposeError(
                f"slices[{i}].insights must be an array"
            )
        if not isinstance(qs, list):
            raise TwinSubstrateCrossAssetMergeComposeError(
                f"slices[{i}].questions must be an array"
            )
        for j, v in enumerate(ins):
            insights.append(
                _require_nonempty(v, field=f"slices[{i}].insights[{j}]")
            )
        for j, v in enumerate(qs):
            questions.append(
                _require_nonempty(v, field=f"slices[{i}].questions[{j}]")
            )

    parent_count = len(parent_asset_ids)
    insight_count = len(insights)
    question_count = len(questions)
    notes.append(
        f"parents={parent_count} · insights={insight_count} · questions={question_count}"
    )

    has_substrate = insight_count + question_count >= 1
    merge_ready = operator_ack and parent_count >= 2 and has_substrate
    if not operator_ack:
        notes.append("merge_ready=false — operator_ack required")
    elif not has_substrate:
        notes.append(
            "merge_ready=false — no insights/questions (no invent substrate)"
        )
    else:
        notes.append("merge_ready=true — provisional cross-asset twin pack")

    notes.extend(
        ("merge_executed=false", "twin_written=false", "store_mutated=false")
    )

    return TwinSubstrateCrossAssetMergeCompose(
        pack_id=pid,
        parent_asset_ids=tuple(parent_asset_ids),
        parent_count=parent_count,
        insight_count=insight_count,
        question_count=question_count,
        insights=tuple(insights),
        questions=tuple(questions),
        merge_ready=merge_ready,
        merge_executed=False,
        twin_written=False,
        store_mutated=False,
        notes=tuple(notes),
        authority="twin_substrate_cross_asset_merge_compose_advisory",
    )


__all__ = [
    "TwinSubstrateCrossAssetMergeCompose",
    "TwinSubstrateCrossAssetMergeComposeError",
    "compose_twin_substrate_cross_asset_merge",
]
